[lldb-dap] Implement `runInTerminal` for Windows

SuibianP opened the pull request:
Currently, the named pipe is passed by name and a transient `ofstream` is constructed at each I/O request. This assumes,
  - Blocking semantics: FIFO I/O waits for the other side to connect.
  - Buffered semantics: Closing one side does not discard existing data.

The former can be replaced by `WaitNamedPipe`/`ConnectNamedPipe` on Win32, but the second cannot be easily worked around. It is also impossible to have another "keep-alive" pipe server instance, as server-client pairs are fixed on connection on Win32 and the client may get connected to it instead of the real one.

Refactor `FifoFile[IO]` to use an open file handles rather than file name.

SuibianP commented:
The patch currently suffers from a critical issue that I cannot diagnose. Help is much appreciated.

When the debug adapter attempts to attach to the launcher,

https://github.com/llvm/llvm-project/blob/4c8c18c8c54308b945d6c78e2c71ebabf29cb554/lldb/tools/lldb-dap/lldb-dap.cpp#L2039

it blocks at

https://github.com/llvm/llvm-project/blob/4c8c18c8c54308b945d6c78e2c71ebabf29cb554/lldb/source/Plugins/Process/Windows/Common/ProcessDebugger.cpp#L580-L581

although adding the following snippet at the front of `RunInTerminalLauncherCommChannel::WaitUntilDebugAdaptorAttaches` indicates that the launcher process does get paused and never resumed.

```cpp
for (int i = 0; !::IsDebuggerPresent(); ++i)
  LLVM_LOGV(log, "Waiting to be attached... {0}", i);
LLVM_LOGV(log, "Attached");
```

SuibianP commented:
Ping

SuibianP commented:
Hi @JDevlieghere, sorry for pinging again but could you please kindly help take a look at this when you have time? Thanks in advance!

JDevlieghere commented:
Sorry, I didn't get the ping (I think because this is marked as a draft). I'm not familiar enough with Windows to review this, so I'm adding some folks that are or work on related stuff in `lldb-dap`.

ashgti commented:
Is the process stopped? I think lldb is waiting for the process to be in a stopped state. In posix terms a `SIGSTOP`, although I'm not as familiar with the win32 equiv (DebugActiveProcess maybe).

SuibianP commented:
As it turns out that Win32 `execvp` actually [creates a new process instead of replace the current process image](https://github.com/python/cpython/issues/101191#issuecomment-1399013437), I changed the implementation to use `CreateProcess` before reporting the PID of the target.

The current implementation works with dape under emacs and should be functional, so I marked it ready for review. On VSCode however, while the extension is able to attach to the process and obtain sort of a stack backtrace, all breakpoints show up as unverified and does not trigger even if re-toggled. I am looking into it now.

---

This is not relevant anymore but for the sake of completion: In the old implementation (wait for attach then `execvp`), the launcher process does get resumed after some 3 minutes. I still do not understand how it happened.

omjavaid commented on lldb/tools/lldb-dap/FifoFiles.h:
```diff
@@ -71,12 +74,14 @@ class FifoFileIO {
-      std::chrono::milliseconds timeout = std::chrono::milliseconds(20000));
+  llvm::Error
+  SendJSON(const llvm::json::Value &json,
+           std::chrono::milliseconds timeout = std::chrono::milliseconds(2000));
```
Is it ok to change to timeout here 20000 --> 2000?

walter-erquinigo replied:
Don't change this. Some targets take a while to launch

SuibianP replied:
Debug leftover; rectified.

omjavaid commented on lldb/tools/lldb-dap/FifoFiles.h:
```diff
@@ -71,12 +74,14 @@ class FifoFileIO {
 
-private:
-  std::string m_fifo_file;
+  // private:
```
Why commenting out private?

SuibianP replied:
Debug leftover; rectified.

omjavaid reviewed (changes requested):
This should be able make lldb\test\API\tools\lldb-dap\runInTerminal\TestDAP_runInTerminal.py pass all tests. Please tests and remove the skipping decorators.

walter-erquinigo commented on lldb/tools/lldb-dap/FifoFiles.h:
```diff
@@ -21,21 +23,22 @@ namespace lldb_dap {
-///     A \a std::shared_ptr<FifoFile> if the file could be created, or an
-///     \a llvm::Error in case of failures.
-llvm::Expected<std::shared_ptr<FifoFile>> CreateFifoFile(llvm::StringRef path);
+std::error_code createNamedPipe(const llvm::Twine &Prefix,
```
This API is unnecessarily complex. The Prefix can just be a StringRef. Also use lower case for names. You also never use the Suffix.
Finally, please write some documentation mentioning that a unique named pipe with the given prefix will be created.
You may consider renaming this function as `createUniqueNamedPipe`

SuibianP replied:
I was mimicking the `llvm::sys::fs::createTemporaryFile` API with the idea that this abstraction might be melded in there at some later time.

Is this considered too broad?

SuibianP replied:
Added doxygen comments, changed to snake_case and renamed function to `createUniqueNamedPipe`.

walter-erquinigo commented on lldb/tools/lldb-dap/RunInTerminal.cpp:
```diff
@@ -111,8 +111,10 @@ Error RunInTerminalLauncherCommChannel::WaitUntilDebugAdaptorAttaches(
-Error RunInTerminalLauncherCommChannel::NotifyPid() {
-  return m_io.SendJSON(RunInTerminalMessagePid(getpid()).ToJSON());
+Error RunInTerminalLauncherCommChannel::NotifyPid(lldb::pid_t pid) {
+  if (pid == 0)
+    pid = getpid();
```
can you explain here why you added this?

SuibianP replied:
I used this as kind of a dynamic default argument. The original codepath always used `getpid()` as the PID to be sent through the pipe, but this does not work for Windows where the PID of the target differs from that of the launcher.

https://github.com/llvm/llvm-project/pull/121269/files#diff-2de0ddf6c080497da43990b05d21e6229d309f51cf269ba7e3860a677f507905R4903-R4905

Maybe an explicit argument without such implicit default would be more readable?

SuibianP replied:
Refactored with explicitly specified arguments.

walter-erquinigo commented on lldb/tools/lldb-dap/RunInTerminal.cpp:
```diff
@@ -158,13 +164,24 @@ std::string RunInTerminalDebugAdapterCommChannel::GetLauncherError() {
+
+  if (cf == NULL)
+    return createStringError(std::error_code(errno, std::generic_category()),
+                             "Error converting file descriptor to C FILE");
```
Please mention runInTerminal here, as otherwise the error is not actionable at all

walter-erquinigo reviewed (commented):
To be honest it's a bit hard to review this because I'm not versed in these low level APIs, let along Windows. Could you make sure that your changes pass mac, linux and windows CI? IIRC you can trigger manually buildbots if you want to run custom tests.

SuibianP commented:
I [enabled `runInTerminal` tests for Windows](https://github.com/llvm/llvm-project/pull/121269/files#diff-38cf10e0a83515e5c6acf5449a3f5de1d63f116d932a32ba169b746826ead1d4) (will fix line 93 later) except for the 4 launcher tests that requires `os.mkfifo` (https://github.com/python/cpython/issues/103510) and they pass on my local computer. This should be sufficient to guarantee the common functionalities. I can do a simple polyfill using `pywin32` for Windows if the ignored test is a blocker for this PR.

All LLDB tests still pass on Linux CI, so there seems to be no regression https://buildkite.com/llvm-project/github-pull-requests/builds/140240

ashgti commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -2249,6 +2249,11 @@ llvm::Error request_runInTerminal(DAP &dap,
                            }
                          });
 
+  llvm::errs() << "WaitForLauncher\n";
```
I think we can remove this

ashgti commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -2249,6 +2249,11 @@ llvm::Error request_runInTerminal(DAP &dap,
 
+  llvm::errs() << "WaitForLauncher\n";
+  auto err = comm_channel.WaitForLauncher();
+  llvm::errs() << "WaitForLauncher returned\n";
```
I think we can remove this as well

SuibianP commented on lldb/tools/lldb-dap/FifoFiles.cpp:
```diff
@@ -24,41 +30,95 @@ using namespace llvm;
-        std::getline(reader, buffer);
-        if (!buffer.empty())
-          line = buffer;
+        rewind(m_fifo_file.m_file);
```
It seemed to help on Windows, but I did not investigate further, as it is technically also required by ISO C standard (C23 draft N3220 §7.23.5.3.7)

> When a file is opened with update mode (’+’ as the second or third character in the previously
described list of mode argument values), both input and output may be performed on the associated
stream. However, output shall not be directly followed by input without an intervening call to the
`fflush` function or to a file positioning function (`fseek`, `fsetpos`, or `rewind`), and input shall not
be directly followed by output without an intervening call to a file positioning function, unless the
input operation encounters end-of-file. Opening (or creating) a text file with update mode may
instead open (or create) a binary stream in some implementations.

ashgti replied:
I think the flow of the logic here is:

* `lldb-dap` (pid 1) opens the fifo
* pid 1 sends sends the `runInTerminal` command.
* pid 1 reads from the fifo (waiting for data)
* `lldb-dap --launch-target` (pid 2) starts
* pid 2 opens the fifo
* pid 2 spawns the new process
* pid 2 writes to the fifo (something like `{"pid":2}\n`) 
* pid 1 finishes the initial read from the fifo
* pid 1 lldb attaches to the process started by pid 2

This should mean that the parent side only reads from the file and the child side only writes to the file. Additionally, I think they're only each performing 1 write/read on each side so their respective file positions should still be at the beginning of the file after opening.

SuibianP replied:
The DA side first sends a `RunInTerminalMessageDidAttach` message in `RunInTerminalDebugAdapterCommChannel::NotifyDidAttach` when it attaches to the launcher. This is also why the FIFO must be `PIPE_ACCESS_DUPLEX`.

ashgti replied:
I think seeking isn't support on a fifo (`lseek` would return `ESPIPE`, see `man 2 lseek` or the Errors section of https://man7.org/linux/man-pages/man2/lseek.2.html). So I'm not sure if you need to do these `rewind` on non-windows platforms.

SuibianP replied:
Practically it should not be needed, but I am tempted to leave the `rewind` there. It helps with standard compliance and does not harm anyways. There might as well be other non-Windows platforms requiring this invocation to properly function.

SuibianP commented on lldb/tools/lldb-dap/FifoFiles.cpp:
```diff
@@ -24,41 +30,95 @@ using namespace llvm;
+#ifdef _WIN32
+      "\\\\.\\pipe\\LOCAL\\"
+#else
+      "/tmp/"
```
On Windows FIFOs live on a virtual hierarchy `\\.\pipe\` not backed a real filesystem. I don't think it is possible to `mkdir` on them, although it is possible to specify something like `\\.\pipe\tmp\dir\myfifo` to `CreateNamedPipe`.

ashgti replied:
`/tmp` is not always the appropriate platform path for non-windows platforms. Thats my main concern.

SuibianP replied:
Hmm.. maybe `llvm::sys::path::system_temp_directory` then?

SuibianP replied:
Changed back to `getPotentiallyUniqueTempFileName` for non-Windows platforms.

SuibianP commented on lldb/tools/lldb-dap/FifoFiles.cpp:
```diff
@@ -9,7 +9,13 @@
 #include "FifoFiles.h"
 #include "JSONUtils.h"
 
```
Removed one empty line and preserved those around `ifdef` guarded includes, the same way `lldb-dap.cpp` does it.

SuibianP commented on lldb/tools/lldb-dap/FifoFiles.cpp:
```diff
@@ -24,41 +30,95 @@ using namespace llvm;
+      path, PIPE_ACCESS_DUPLEX | FILE_FLAG_FIRST_PIPE_INSTANCE,
+      PIPE_TYPE_BYTE | PIPE_READMODE_BYTE | PIPE_WAIT, 1, 1024, 1024, 0, NULL);
+  if (h == INVALID_HANDLE_VALUE)
+    return std::error_code(::GetLastError(), std::system_category());
```
Didn't know this utility function existed. Switched to it.

SuibianP commented on lldb/tools/lldb-dap/FifoFiles.cpp:
```diff
@@ -24,41 +30,95 @@ using namespace llvm;
+    return std::error_code(::GetLastError(), std::system_category());
+  ResultFd = _open_osfhandle((intptr_t)h, _O_TEXT | _O_RDWR);
+  if (ResultFd == -1)
+    return std::error_code(::GetLastError(), std::system_category());
```
Switched.

SuibianP commented on lldb/tools/lldb-dap/FifoFiles.cpp:
```diff
@@ -24,41 +30,95 @@ using namespace llvm;
+FifoFile::FifoFile(StringRef path, FILE *f) : m_path(path), m_file(f) {}
+FifoFile::FifoFile(FifoFile &&other)
+    : m_path(other.m_path), m_file(other.m_file) {
+  other.m_file = nullptr;
```
Called `clear` on it.

Also fixed one place where the moved `m_path` was still used.

SuibianP commented on lldb/tools/lldb-dap/FifoFiles.h:
```diff
@@ -21,21 +23,22 @@ namespace lldb_dap {
 struct FifoFile {
   FifoFile(llvm::StringRef path);
+  FifoFile(llvm::StringRef path, FILE *f);
+  // FifoFile(llvm::StringRef path, FILE *f);
```
Removed.

SuibianP commented on lldb/tools/lldb-dap/FifoFiles.cpp:
```diff
@@ -24,41 +30,95 @@ using namespace llvm;
 namespace lldb_dap {
 
-FifoFile::FifoFile(StringRef path) : m_path(path) {}
+std::error_code EC;
```
Unintended, moved into function.

ashgti commented on lldb/tools/lldb-dap/FifoFiles.h:
```diff
@@ -21,21 +23,28 @@ namespace lldb_dap {
-llvm::Expected<std::shared_ptr<FifoFile>> CreateFifoFile(llvm::StringRef path);
+/// Note that the resulting filename is further prepended by \c \\.\pipe\\LOCAL\
+/// on Windows and \c /tmp on other platforms.
+std::error_code createUniqueNamedPipe(const llvm::Twine &prefix,
```
Should this return an `llvm::Error` instead of a `std::error_code`? `llvm::Error` has stronger requirements for handling errors.

SuibianP replied:
Changed.

ashgti commented on lldb/tools/lldb-dap/RunInTerminal.cpp:
```diff
@@ -158,13 +156,25 @@ std::string RunInTerminalDebugAdapterCommChannel::GetLauncherError() {
+  if (setvbuf(cf, NULL, _IONBF, 0))
+    return createStringError(std::error_code(errno, std::generic_category()),
+                             "Error setting unbuffered mode on C FILE");
+  // There is no portable way to conjure an ofstream from HANDLE, so use FILE *
+  // llvm::raw_fd_stream does not support getline() and there is no
+  // llvm::buffer_istream
```
If you have the path, can't you use that to open the file with `std::ofstream`?

SuibianP replied:
On the client (launcher) side it should be possible, but on the server end I suppose not. Windows has the distinction of "server" (from `CreateNamedPipe`) and "client" (from `CreateFile`) ends of FIFOs and unlike on POSIX where FIFOs can read from any previous write, clients can only communicate with server but not fellow clients.

Opening a `std::ofstream` (which underlyingly is `CreateFile`) would open a client of the pipe in the same process that connects to the server. The server side must use `CreateNamedPipe` which gives out a `HANDLE`, thus the issue of turning that into some platform independent file object.

SuibianP commented on lldb/tools/lldb-dap/FifoFiles.cpp:
```diff
@@ -24,41 +30,95 @@ using namespace llvm;
+  if (EC)
+    return EC;
 #endif
+  result_path.pop_back();
```
Yes, it is the NUL. It would be encoded into JSON if not popped.

I think it might have always been buggy and just happened to work, as NUL termination from `StringRef::data` is explicitly not guaranteed.

SuibianP commented on lldb/tools/lldb-dap/FifoFiles.cpp:
```diff
@@ -24,41 +30,95 @@ using namespace llvm;
-          line = buffer;
+        rewind(m_fifo_file.m_file);
+        constexpr size_t buffer_size = 2048;
+        char buffer[buffer_size];
+        char *ptr = fgets(buffer, buffer_size, m_fifo_file.m_file);
+        if (ptr == nullptr || *ptr == 0)
+          return;
+        size_t len = strlen(buffer);
+        if (len <= 1)
+          return;
+        buffer[len - 1] = '\0'; // remove newline
+        line = buffer;
```
We can reinvent POSIX `getline`, but it is probably not worth the effort. Any JSON schema change that overruns the buffer will absolutely get truncated, break the test visibly and get caught during review.

SuibianP commented on lldb/tools/lldb-dap/FifoFiles.cpp:
```diff
@@ -24,41 +30,95 @@ using namespace llvm;
-        if (!buffer.empty())
-          line = buffer;
+        rewind(m_fifo_file.m_file);
+        constexpr size_t buffer_size = 2048;
```
The two sizes are not related. The one passed to `CreateNamedPipe` is the kernel buffer, while this one is for the receiving user buffer. I do need to rethink how long the buffer needs to be, though.

SuibianP replied:
Fixed at 512 which should be longer than any single message that passes through the FIFO. Mentioned in code comment as well.

SuibianP committed:
[lldb-dap] Implement runInTerminal for Windows

Currently, the named pipe is passed by name and a transient ofstream is
constructed at each I/O request. This assumes,
  - Blocking semantics: FIFO I/O waits for the other side to connect.
  - Buffered semantics: Closing one side does not discard existing data.

The former can be replaced by WaitNamedPipe/ConnectNamedPipe on Win32,
but the second cannot be easily worked around. It is also impossible to
have another "keep-alive" pipe server instance, as server-client pairs
are fixed on connection on Win32 and the client may get connected to it
instead of the real one.

Refactor FifoFile[IO] to use an open file handles rather than file name.

---

Win32 provides no way to replace the process image. Under the hood exec*
actually creates a new process with a new PID. DebugActiveProcess also
cannot get notified of process creations.

Create the new process in a suspended state and resume it after attach.

SuibianP commented on lldb/tools/lldb-dap/FifoFiles.cpp:
```diff
@@ -24,41 +30,95 @@ using namespace llvm;
-#if defined(_WIN32)
-  return createStringError(inconvertibleErrorCode(), "Unimplemented");
+// This probably belongs to llvm::sys::fs as another FSEntity type
+std::error_code createNamedPipe(const Twine &Prefix, StringRef Suffix,
```
Changed to return `llvm::Error`.

SuibianP commented on lldb/tools/lldb-dap/FifoFiles.cpp:
```diff
@@ -24,41 +30,95 @@ using namespace llvm;
+    EC = std::error_code(errno, std::generic_category());
+    llvm::errs() << "Failed to open fifo file " << path << ": " << EC.message()
+                 << "\n";
+    std::terminate();
```
Change to static factory method `FifoFile::create` that returns an `Expected<FifoFile>`.

SuibianP commented on lldb/test/API/tools/lldb-dap/runInTerminal/main.c:
```diff
@@ -1,11 +1,13 @@
 #include <stdlib.h>
-#include <unistd.h>
+
+#include <threads.h>
```
[`threads.h`](https://en.cppreference.com/w/c/header/threads) is in ISO C11 standard §7.26. The tests are all [passing in Linux x64 CI](https://buildkite.com/llvm-project/github-pull-requests/builds/151090#0195426e-ca2c-4da7-b72b-6ac347c97b90/429-1562).

SuibianP replied:
Turns out Apple does not yet support C11 `threads.h`.

https://github.com/llvm/llvm-project/blob/7e8a06cfa4a2951b8ee77e19e34926e6e535b4d1/clang/lib/Basic/Targets/OSTargets.cpp#L61-L62

Changed test case to use C++11 `thread` which is supported.

SuibianP commented on lldb/tools/lldb-dap/FifoFiles.cpp:
```diff
@@ -24,41 +30,95 @@ using namespace llvm;
+    std::terminate();
+  }
+  if (setvbuf(m_file, NULL, _IONBF, 0))
+    llvm::errs() << "Error setting unbuffered mode on C FILE\n";
```
Changed.

SuibianP commented on lldb/tools/lldb-dap/FifoFiles.cpp:
```diff
@@ -78,8 +143,12 @@ Error FifoFileIO::SendJSON(const json::Value &json,
       new std::future<void>(std::async(std::launch::async, [&]() {
-        std::ofstream writer(m_fifo_file, std::ofstream::out);
-        writer << JSONToString(json) << std::endl;
+        rewind(m_fifo_file.m_file);
```
For future reference: refer to the comment chain below.

SuibianP commented:
Pinging @ashgti and @walter-erquinigo for review before I forget again

SuibianP commented:
Pinging again for review @ashgti @omjavaid @walter-erquinigo 

Also want to highlight https://github.com/SuibianP/llvm-project/commit/f9675fbe23abe59c3cf56a6263b7a78ec20e42d4#diff-ed319b24adfa654beeff8b32f9d8f975c62299fbe5cb6fbe1028aa6bbaa369e7. Not sure how it is passing CI though; I might as well be understanding it wrongly.

SuibianP commented on lldb/tools/lldb-dap/FifoFiles.cpp:
```diff
@@ -24,41 +31,105 @@ using namespace llvm;
+  const char *path = result_path.data();
+#ifdef _WIN32
+  HANDLE h = ::CreateNamedPipeA(
+      path, PIPE_ACCESS_DUPLEX | FILE_FLAG_FIRST_PIPE_INSTANCE,
```
> when its opened for reading it blocks until the file is also opened for writing and vice versa

This is implemented with [`ConnectNamedPipe`](https://learn.microsoft.com/en-us/windows/win32/api/namedpipeapi/nf-namedpipeapi-connectnamedpipe) in `FifoFileIO::WaitForPeer` after creation. `WaitNamedPipe` was not needed as the server side would already be around by the time the client tries to connect.

> deadlock the process by having it wait on itself accidentally

Intra-process deadlock (waiting on its own pending output) should not be possible, since on Windows a process cannot even read back what it wrote into a named pipe from the same handle. The data flow is bidirectional, but can only be consumed by the other end.

Inter-process deadlock, with the exception of buffering, is also rather unlikely. Current protocol over the pipe is strictly synchronous and sequential, or "half-duplex" in some sense -- there are no overlapping transmissions; in other words, the process will not transmit before expected response is fully received, and will not block on read before it sends out the message.

Buffering can still cause deadlocks. Without the `rewind`, the process proceeds to wait for the response after write, but the payload has somehow not fully arrived at the other end, and the peer as a result does not respond yet. I speculated it to be due to libc buffering, but did not manage to penetrate through MS CRT code to verify.

> Should we instead switch to a communication channel that supports reading and writing for all platforms?

Using sockets feels a bit like overkill, but would definitely make code more uniform across platforms I suppose. I can give it a try if you deem it more maintainable.

> we have some existing helpers for supporting sockets already.

As an aside, I remember I did also attempt to retrofit the use case into `lldb_private::Pipe`, but gave up because duplex was too hard to shoehorn into the existing abstraction, and would have very limited utility even if implemented.

> Is that something you might have a chance to look into?

I am not too familiar with that so unable to give guarantees, but let me give it a try.
Also unsure if the socket rewrite is going to make it for LLVM 21 feature freeze in July.
